### PR TITLE
RL4J - Fix for setTarget() (issue #8107)

### PR DESCRIPTION
**Pull request by aboulang2002**
## What changes were proposed in this pull request?

See issue #8107 

After reading again "_Playing Atari with Deep Reinforcement Learning_" (http://arxiv.org/abs/1312.5602), I still think there is a problem with the current implementation of standard DQN, but I don't think that the target-network should be used for `dqnOutputAr`.

I have also read the section 3 "_Understanding Deep Q-Network_" of "A _Theoretical Analysis of Deep Q-Learning_" (https://arxiv.org/pdf/1901.00137.pdf).
And also https://medium.com/@qempsil0914/deep-q-learning-part2-double-deep-q-network-double-dqn-b8fc9212bbb2 which has a good overview of DQN.
Plus I reviewed a few DQN implementations on github. For example, Google's TensorLayer (https://github.com/tensorlayer/tensorlayer/blob/72666dbf23718e8710dc1069dbcf3231d131dff2/examples/reinforcement_learning/tutorial_DQN_variants.py).

First, the target-network should be used to compute `tempQ` in the standard DQN algorithm.
This can be seen in the equation 3.1 (and the one above in the text) of https://arxiv.org/pdf/1901.00137.pdf and also from lines 6-8 of `tutorial_DQN_variants.py`:
```
	The max operator in standard DQN uses the same values both to select and to evaluate an action by
	Q(s_t, a_t) = R_{t+1} + \gamma * max_{a}Q_{tar}(s_{t+1}, a)
```
We can see that the target-network (`Q_{tar}`) should be used.

Second, the target-network is only used as a stable base to compute the loss function, and not "to obtain the labels" as @flyheroljg says. This means that the Q-Network should used for `dqnOutputAr`. This can be seen on line 8 of `tutorial_DQN_variants.py` and also on lines 322-325:
```
                # calculate loss
                with tf.GradientTape() as q_tape:
                    b_q = tf.reduce_sum(qnet(b_o) * tf.one_hot(b_a, out_dim), 1)
                    loss = tf.reduce_mean(huber_loss(b_q - (b_r + reward_gamma * b_q_)))
```
The above snippet is used for Double-DQN, but only the td-target part (`(b_r + reward_gamma * b_q_)` above) is different between standard and double DQN in the computation of the loss function.
This can also be seen in the section "_Double Q-Learning Algorithm_" of https://medium.com/@qempsil0914/deep-q-learning-part2-double-deep-q-network-double-dqn-b8fc9212bbb2, as well as equation 3.3 of https://arxiv.org/pdf/1901.00137.pdf

## How was this patch tested?

With cartpole-v0 (re-written in java because couldn't make gym's version to work; can add it to this PR if requested)

### CODE REVIEW

**Review by saudet — APPROVED** [review]
That looks good! Please do add your version of Cartpole, and let's merge this :) Thanks

**Comment by aboulang2002** [comment]
I added CartpoleNative, a native java port of GYM's cartpole-v0. The rendering has not been ported. I may add it later but it's lower on my list of todos.
